The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3380435120/jobs/5613214705) [(plain)](https://github.com/rust-lang/rust/commit/53f5bb385289914afe1c66b5012e7101c67647ee/checks/9257656380/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 36f8da3a32ea
Successfully tagged rust-ci:latest
Built container sha256:36f8da3a32ea9d041790db9f4ecadb41c90ad438c8881328d1cc808483d01c19
Uploading finished image to https://ci-caches.rust-lang.org/docker/5df1423b252af8fe6e1c1d4bfe9ecee415e5079d069fe81f116b5415ff8f3fff1d8d7f6f046a60e231d4eebdb0071aabb1a1f0159ce874f5e50913c8f8929e6e
upload failed: - to s3://rust-lang-ci-sccache2/docker/5df1423b252af8fe6e1c1d4bfe9ecee415e5079d069fe81f116b5415ff8f3fff1d8d7f6f046a60e231d4eebdb0071aabb1a1f0159ce874f5e50913c8f8929e6e Unable to locate credentials
[CI_JOB_NAME=x86_64-gnu-llvm-13]
---
failures:

---- [assembly] src/test/assembly/asm/aarch64-types.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-types/aarch64-types.s" "/checkout/src/test/assembly/asm/aarch64-types.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/assembly/asm/aarch64-types.rs:100:11: error: CHECK: expected string not found in input
// CHECK: str {{.*}}x30
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-types/aarch64-types.s:35:13: note: scanning from here
issue_75761:
            ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-types/aarch64-types.s:37:2: note: possible intended match here
/checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-types/aarch64-types.s:37:2: note: possible intended match here
 str x28, [sp, #-16]!

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/assembly/asm/aarch64-types/aarch64-types.s
Check file: /checkout/src/test/assembly/asm/aarch64-types.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
             .
             .
             .
             .
            30:  
            31:  .section .text.issue_75761,"ax",@progbits 
            32:  .globl issue_75761 
            33:  .p2align 2 
            34:  .type issue_75761,@function 
            35: issue_75761: 
check:100'0                 X error: no match found
            36:  .cfi_startproc 
check:100'0     ~~~~~~~~~~~~~~~~
            37:  str x28, [sp, #-16]! 
check:100'0     ~~~~~~~~~~~~~~~~~~~~~~
check:100'1      ?                     possible intended match
            38:  .cfi_def_cfa_offset 16 
check:100'0     ~~~~~~~~~~~~~~~~~~~~~~~~
            39:  .cfi_offset w28, -16 
check:100'0     ~~~~~~~~~~~~~~~~~~~~~~
            40:  //APP 
check:100'0     ~~~~~~~
            41:  //NO_APP 
check:100'0     ~~~~~~~~~~
            42:  ldr x28, [sp], #16 
check:100'0     ~~~~~~~~~~~~~~~~~~~~
             .
             .
>>>>>>
------------------------------------------
```

</details>

--- a/compiler/rustc_codegen_llvm/src/asm.rs
+++ b/compiler/rustc_codegen_llvm/src/asm.rs
@@ -496,6 +496,44 @@ fn xmm_reg_index(reg: InlineAsmReg) -> Option<u32> {
     }
 }
 
+/// If the register is an AArch64 integer register then return its index.
+fn a64_reg_index(reg: InlineAsmReg) -> Option<u32> {
+    match reg {
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x0) => Some(0),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x1) => Some(1),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x2) => Some(2),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x3) => Some(3),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x4) => Some(4),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x5) => Some(5),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x6) => Some(6),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x7) => Some(7),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x8) => Some(8),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x9) => Some(9),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x10) => Some(10),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x11) => Some(11),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x12) => Some(12),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x13) => Some(13),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x14) => Some(14),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x15) => Some(15),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x16) => Some(16),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x17) => Some(17),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x18) => Some(18),
+        // x19 is reserved
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x20) => Some(20),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x21) => Some(21),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x22) => Some(22),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x23) => Some(23),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x24) => Some(24),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x25) => Some(25),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x26) => Some(26),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x27) => Some(27),
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x28) => Some(28),
+        // x29 is reserved
+        InlineAsmReg::AArch64(AArch64InlineAsmReg::x30) => Some(30),
+        _ => None,
+    }
+}
+
 /// If the register is an AArch64 vector register then return its index.
 fn a64_vreg_index(reg: InlineAsmReg) -> Option<u32> {
     match reg {
@@ -526,6 +564,22 @@ fn reg_to_llvm(reg: InlineAsmRegOrRegClass, layout: Option<&TyAndLayout<'_>>) ->
                     'x'
                 };
                 format!("{{{}mm{}}}", class, idx)
+            } else if let Some(idx) = a64_reg_index(reg) {
+                let class = if let Some(layout) = layout {
+                    match layout.size.bytes() {
+                        8 => 'x',
+                        _ => 'w',
+                    }
+                } else {
+                    // We use i32 as the type for discarded outputs
+                    'w'
+                };
+                if class == 'x' && reg == InlineAsmReg::AArch64(AArch64InlineAsmReg::x30) {
+                    // LLVM doesn't recognize x30. use lr instead.
+                    "{lr}".to_string()
+                } else {
+                    format!("{{{}{}}}", class, idx)
+                }
             } else if let Some(idx) = a64_vreg_index(reg) {
                 let class = if let Some(layout) = layout {
                     match layout.size.bytes() {
@@ -541,9 +595,6 @@ fn reg_to_llvm(reg: InlineAsmRegOrRegClass, layout: Option<&TyAndLayout<'_>>) ->
                     'q'
                 };
                 format!("{{{}{}}}", class, idx)
-            } else if reg == InlineAsmReg::AArch64(AArch64InlineAsmReg::x30) {
-                // LLVM doesn't recognize x30
-                "{lr}".to_string()
             } else if reg == InlineAsmReg::Arm(ArmInlineAsmReg::r14) {
                 // LLVM doesn't recognize r14
                 "{lr}".to_string()

--- a/src/test/ui/asm/aarch64/llvm-58384.rs
+++ b/src/test/ui/asm/aarch64/llvm-58384.rs
@@ -1,0 +1,16 @@
+// only-aarch64
+// run-pass
+// needs-asm-support
+
+// Test that we properly work around this LLVM issue:
+// https://github.com/llvm/llvm-project/issues/58384
+
+use std::arch::asm;
+
+fn main() {
+    let a: i32;
+    unsafe {
+        asm!("", inout("x0") 435 => a);
+    }
+    assert_eq!(a, 435);
+}

